File: Data Structures & Algorithms/subsets/submission-1.java
class Solution {
    int len = 0;
    int[] x;
    List<List<Integer>> ans = new ArrayList<>();
    public List<List<Integer>> subsets(int[] nums) {
        len = nums.length;
        x = nums;
        dfs(0, new ArrayList<Integer>());
        return ans;
    }
    public void dfs(int i, List<Integer> nums) {
        if (i >= len) {
            return;
        }

        dfs(i + 1, nums);
        if (i == len-1 ){
            ans.add(new ArrayList<>(nums));
        }

        nums.add(x[i]);
        dfs(i + 1, nums);
        if (i == len-1 ){
            ans.add(new ArrayList<>(nums));
        }
        nums.removeLast();
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        lenght = len(nums)
        visited = [-1] * (lenght)
        def dfs(n: int) -> int:
            if visited[n] != -1:
                return visited[n]
            print("зашли в дфс", n)
            ans = 1
            for i in range(n+1 , lenght):
                if nums[i] > nums[n]:
                    ans = max(ans , 1+dfs(i))
            visited[n] = ans
            print("---------- вышли дфс", n)
            return ans
        ans = 0
        for i in range(lenght):
            ans = max(ans , dfs(i))
        return ans


